apps/prairielearn/assets/scripts/lib/questionsTable.ts

```diff
@@ -0,0 +1,278 @@
+        title: 'External Grading Image',
+        visible: false,
+        headerFilter: 'list',
+        headerFilterPlaceholder: '(All Images)',
+        headerFilterFunc: (headerValue: string, rowValue: string) =>
+          headerValue === '(No Image)' ? !rowValue : headerValue === rowValue,
+        headerFilterParams: {
+          values: [
+            { label: '(All Images)' },
+            ...uniq(questions.map((q) => q.external_grading_image || '(No Image)')).sort(),
+          ],
+        },
+      },
+      ...course_instances.map((ci) => ({
+        field: `assessments_${ci.id}`,
+        title: `${ci.short_name} Assessments`,
+        mutator: (_value: any, data: QuestionsPageDataAnsified): AssessmentsFormatForQuestion =>
+          data.assessments?.filter((a) => idsEqual(a.course_instance_id, ci.id)) ?? [],
+        visible: ci.current,
+        headerSort: false,
+        formatter: (cell: CellComponent) =>
+          (cell.getValue() as AssessmentsFormatForQuestion)
+            ?.map((assessment) =>
+              html`<a
+                href="${plainUrlPrefix}/course_instance/${ci.id}/instructor/assessment/${assessment.assessment_id}"
+                class="badge color-${assessment.color} color-hover"
+                onclick="event.stopPropagation();"
+              >
+                <span>${assessment.label}</span>
+              </a>`.toString(),
+            )
+            .join(' '),
+        headerFilter: 'list' as const,
+        headerFilterPlaceholder: '(All Assessments)',
+        headerFilterFunc: (headerValue: string, rowValue: AssessmentsFormatForQuestion) =>
+          headerValue === '0'
+            ? !rowValue?.length
+            : rowValue?.some((row) => headerValue === row.label),
+        headerFilterParams: {
+          values: [
+            { label: '(All Assessments)' },
+            { label: '(None)', value: '0' },
+            ...uniq(
+              questions
+                .map(
+                  (q) =>
+                    q.assessments
+                      ?.filter((a) => idsEqual(a.course_instance_id, ci.id))
+                      .map((a) => a.label) ?? [],
+                )
+                .flat(),
+            ).sort(),
+          ],
+        },
+      })),
+    ],
+  });
+
+  selectFilterOnSearch(table);
+
+  table.on('renderComplete', () => {
+    // Popovers must be reloaded when the table is rendered (e.g., after a page change or filter)
+    $('.js-sync-popover[data-toggle="popover"]')
+      .popover({ sanitize: false })
+      .on('show.bs.popover', function () {
+        $($(this).data('bs.popover').getTipElement()).css('max-width', '80%');
+      });
+  });
+
+  document
+    .querySelector<HTMLButtonElement>('.js-clear-filters-btn')
+    ?.addEventListener('click', () => {
+      table.clearFilter(true);
+    });
+});
+
+function qidFormatter(cell: CellComponent): string {
+  const { urlPrefix, qidPrefix } = decodeData<EncodedQuestionsData>('questions-table-data');
```

Comment:
Decoding the _entire_ table's data once per row is horrible for performance. You should decode `questions-table-data` exactly once on page startup and use those values whenever needed.

Reply:
Moved to formatter parameters in 22a24fe4ab3f3b16bd97d48429e340b6dfb887ac. This will store the values in the column definition, which will be passed as argument to the formatter.